### PR TITLE
[SwiftSyntax] Swift Diagnostics API

### PR DESCRIPTION
**Pull request by harlanhaskins**
This patch provides an initial implementation of a diagnostic serialization API from Swift. This allows popping diagnostics from Swift files and serializing them out to JSON.

### CODE REVIEW

**Comment by harlanhaskins** [comment]
I'd like to pick this back up.

@jrose-apple: I'm waffling back and forth between representations here. Should diagnostics be attached to `SourceLocation`s or `Syntax` nodes? Should `SourceLocation` contain a file in it, or should the `DiagnosticEngine` own the file it's emitting diagnostics into? How can we associate a `Syntax` node to its original file?

**Comment by harlanhaskins** [comment]
@nkcsgexi This is the branch I was talking about. I'd like to pick this back up!

**Comment by nkcsgexi** [comment]
Wow, this will allow Swift libSyntax API clients to issue diagnostics for by using Swiftc; is my understanding correct?

**Comment by harlanhaskins** [comment]
With a little work. Essentially, this library as-is contains a diagnostic engine implementation in Swift that can serialize to Clang's LLVM bitstream-based diagnostic format.

I foresee linters being able to produce _real_ diagnostics.

**Comment by harlanhaskins** [comment]
@nkcsgexi So @CodaFi and I explored [a diagnostic design for Silt](https://github.com/silt-lang/silt/blob/master/Sources/Lithosphere/Diagnostic.swift) which I think has a lot of strong suits.

For one, they attach diagnostics to Syntax nodes directly instead of to source locations. Another benefit is that it's centered around `Diagnostic.Message` which is a struct containing text and a severity. To create diagnostics, you add a member onto `Diagnostic.Message` in an extension, then you use that directly. It mirrors the `ERROR("unexpected thing '%0'", (StringRef)))` macros.

```swift
extension Diagnostic.Message {
  static let unexpectedEOF = Diagnostic.Message(.error, "unexpected end-of-file")
  static func unexpectedToken(_ token: TokenSyntax) -> Diagnostic.Message {
    return .init(.error, "unexpected token '\(token.text)'")
  }
}
```

Then you just call

```
engine.diagnose(.unexpectedEOF, node: lastToken)
```

If you want to add highlights or notes, you can use the special `builder initializer,` similar to the Syntax nodes' builder initializer.

```
extension Diagnostic.Message {
  static func expected(_ kind: TokenKind) -> Diagnostic.Message {
    return .init(.note, "expected '\(kind.text)'")
  }
}
engine.diagnose(.unexpected(token), node: token) { diag in
  diag.highlight(token)
  diag.note(.expected(.semicolon)), node: token)
}
```

It gives all the features we have now in C++ with the RAII-based `InFlightDiagnostic` with a much safer, more idiomatically Swift interface.

**Comment by nkcsgexi** [comment]
@harlanhaskins I like the idea of attaching diagnostics to libSytnax nodes directly as API design. The only concern is we are marrying two things of different life spans. libSyntax nodes should outlive any diagnostics; and diagnostics shouldn't exist after we handle them once. Do we have a good story about controlling the lifespans of these two things on client side?

**Comment by harlanhaskins** [comment]
I think it should be the diagnostic engine's job to map Syntax nodes to their SourceLocations. At the end of the day, clang-style diagnostics are SourceLocation-based, and we can't quite change that.

So I think if the DiagnosticEngine received nodes as API, then did clever things to find the original source location of the beginning of a node (after skipping leading trivia), then we'd have a clear distinction between lifetimes.

File: tools/SwiftSyntax/Diagnostic.swift [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+  public let file: String
+}
+
+/// Represents a start and end location in a Swift file.
+public struct SourceRange {
+  /// The beginning location in the source range.
+  public let start: SourceLocation
+
+  /// The ending location in the source range.
+  public let end: SourceLocation
```
Comment (jrose-apple) [inline tools/SwiftSyntax/Diagnostic.swift]: Do we ever need ending line/column/file information for ranges, or would a "length" field suffice?

File: tools/SwiftSyntax/BitCodeAbbrev.swift [hunk 2]
```diff
@@ -0,0 +1,55 @@
+
+import Foundation
+
+typealias BitCodeAbbrev = [BitCodeAbbrevOp]
+
+enum BitCodeAbbrevOp {
```
Comment (jrose-apple) [inline tools/SwiftSyntax/BitCodeAbbrev.swift]: Oh geez. Well then.

File: tools/SwiftSyntax/Diagnostic.swift (hunk 1 above)
Comment (harlanhaskins) [reply tools/SwiftSyntax/Diagnostic.swift]: Would we lose multi-line ranges?

Comment (harlanhaskins) [reply tools/SwiftSyntax/Diagnostic.swift]: Granted, clang and Swift don't really handle multi-line ranges very well (at all?)

File: tools/SwiftSyntax/BitCodeAbbrev.swift (hunk 2 above)
Comment (harlanhaskins) [reply tools/SwiftSyntax/BitCodeAbbrev.swift]: Should this be a clearer name?

**Comment by jrose-apple** [comment]
Some downsides of attaching diagnostics to Syntax nodes:
- Need to be very careful when diagnosing a transformed AST; where the diagnostic ends up being printed may not be obvious
- Still need to be able to do sub-token-level diagnostics sometimes (usually for fix-its, though, not where to put the diagnostic)
- Notes often go on imported declarations, which have neither Syntax nodes nor source locations

But I think all that means is that there should be *some* way to use a source location or something else to emit a diagnostic, not that it has to be the primary mechanism. Most of the time in Swift we emit diagnostics on declarations when we can, and this seems likes a logical extension of that.

**Comment by harlanhaskins** [comment]
Yeah, diagnosing transformations kinda throws a wrench into this whole thing. Deriving source locations from a transformed AST is also almost certainly going to give a bad result. A previous design of this made the `DiagnosticEngine` keep track of a single `SourceFileSyntax,` where diagnostics emitted on a node with a different root would cause an error at runtime -- maybe this is something we should keep?

Also, another glaring issue: Where do I get the filename if I just use a Syntax node? That isn't stored anywhere in the node hierarchy. 🙃

File: tools/SwiftSyntax/Diagnostic.swift (hunk 1 above)
Comment (jrose-apple) [reply tools/SwiftSyntax/Diagnostic.swift]: I assumed it would be length-in-bytes, which is fine for a fix-it and probably sufficient for a highlight.

Comment (jrose-apple) [reply tools/SwiftSyntax/Diagnostic.swift]: Clang and Swift both handle multi-line ranges fine as long as you're not using textual diagnostic output.

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

File: tools/SwiftSyntax/Diagnostic.swift (hunk 1 above)
Comment (harlanhaskins) [reply tools/SwiftSyntax/Diagnostic.swift]: Seems we do need an end location, at least for highlights in the `.dia` format.

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

File: tools/SwiftSyntax/SerializedDiagnosticConsumer.swift [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+// This file provides an implementation of the Clang diagnostic serialization
+// format for diagnostics that have been added to a DiagnosticEngine.
+//===----------------------------------------------------------------------===//
```
Comment (nkcsgexi) [inline tools/SwiftSyntax/SerializedDiagnosticConsumer.swift]: The above doc-comment seems not right.

File: tools/SwiftSyntax/SerializedDiagnosticConsumer.swift [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+      case .sourceRange: return "SrcRange"
+      case .diagnosticFlag: return "DiagFlag"
+      case .category: return "CatName"
+      case .filename: return "FileName"
+      case .fixIt: return "FixIt"
+      }
+    }
+  }
+
+  let stream = BitstreamWriter()
```
Comment (nkcsgexi) [inline tools/SwiftSyntax/SerializedDiagnosticConsumer.swift]: Do we need an application-specific bit stream writer in swiftc to actually emit these diags?

**Comment by akyrtzi** [comment]
> Essentially, this library as-is contains a diagnostic engine implementation in Swift that can serialize to Clang's LLVM bitstream-based diagnostic format.
>
>I foresee linters being able to produce real diagnostics.

Hey Harlan, I have a high-level comment on this. Nothing is really concrete but we have some initial thoughts and we are considering that a linter tool could be integrated as part of building by having the swift driver invoke the tool and then *merge* the diagnostics of the linter tool with the diagnostics of the compiler (essentially write them out to the same .dia file that the compiler diagnostics will be written).

The interaction with the tool and the driver could be through stdin/stdout and using json as communication mechanism. Essentially sent some json for the information the tool would need (e.g the libSyntax tree) and receive some json from the tool, e.g. the diagnostics. In that model, could we adapt the diagnostics engine on the SwiftSyntax side to also have a json format serializer ?

Note that in such a model we may not end up using the bitstream writer. It seems to me there's not much benefit to write them out to a separate .dia file for the driver to read, if the tool could just pass them along as json.

What do you think ?

**Comment by harlanhaskins** [comment]
I’m down for pretty much anything here, as long as there’s a good interchange format. My main concern with using stdin/stdout is that it I don’t want these tools to be _only_ diagnostic-oriented.

**Comment by harlanhaskins** [comment]
Like I could foresee static analysis tools, code generators, test discovery tools, documentation generators, etc. that might want to customize their output beyond just diagnostics. Though I suppose they wouldn’t be invoked by `swiftc`...

**Comment by akyrtzi** [comment]
The tool could write out any auxiliary data files as it wishes, and could communicate back the path for them, stdin/stdout is not a limiting factor I think.

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
@akyrtzi So the current branch uses the barebones `Codable` implementation to synthesize JSON diagnostics. Ideally, we come up with a schema we like rather than the default, I just wanted to show it's pretty simple to use JSON in Swift at least. Teaching swiftc to convert it would be more difficult, but definitely doable.

**Comment by nkcsgexi** [comment]
@harlanhaskins do you think this is ready to be merged? or there're some remaining work to be done.

**Comment by harlanhaskins** [comment]
I’m gonna get rid of `runSwiftTool` because I want to figure it out more. Also there’s no reliable way right now to get source locations from a Syntax node, but we can add that later.

**Comment by nkcsgexi** [comment]
By "getting rid of runSwiftTool", do you mean we'll run these tools via `swiftc`? and in that way, `swiftc` needs to decode these serialized diagnostics and actually emits them, right?

**Comment by harlanhaskins** [comment]
I want to just commit the diagnostic API in this, and incorporate the swiftc running in a separate PR later, where we can talk more about the API structure.

**Comment by nkcsgexi** [comment]
Sounds great! I was thinking about having a way to diagnose any "UnknownXXX" appearing in a given source file. For instance, we run a command `swiftc -lint -lint-checker=checker.swift` where `checker.swift` visits all libSyntax nodes and errs any unknown entities. The output will be error messages indicating the source ranges of any unknown entities. IMO, this will be a good way to diagnose/test libSyntax parsing.

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
Thoughts on merging as-is?

**Review by nkcsgexi — APPROVED** [review]

**Comment by harlanhaskins** [comment]
:boat:

**Pull request merged**